### DISCUSSION TITLE
Performance: retrieve Textures from ContentManager vs from its reference

### DISCUSSION THREAD

**Discussion by KenshirouLuke** [MonoGame/MonoGame#8426] [Content]
Hello,
I was wondering, after the LoadContent phase, if retrieving textures from the ContentManager can be an expensive operation when there are a lot of textures loaded in the same ContentManager.
I'm referring to the following line of code, that loads the asset when called the first time, and retrieves its instance when called the next times:
`Content.Load<Texture2D>(assetName)`

Since this can happen a lot of times in each Update cycle (ie: check for instances to spawn, so retrieve their textures each time for the constructor), I was wondering if storing also locally (ie: in the screen class that contains its ContentManager) all the loaded Texture2D objects that are potentially needed in the Update cycle, can be considered as a best practice, since its retrieval is immediate (direct access to the local reference) insted of getting it each time from the Content's dictionary:
`Dictionary<string, object> loadedAssets`

I know the dictionary is fast since is using a key, but also the key is a string, and potentially a long one (think about the path in addition to the filename), and could be among lot of similar strings (like those that are sharing the path) so the retrieval could not be immediate.

I also discovered that there is only a dictionary for all types of content, and probably it would be faster to have a bunch of them, one for content type, to really speed up the search, instead of only check later that the content type is the desired one?
```
if (loadedAssets.TryGetValue(key, out value) && value is T)
{
    return (T)value;
}
```

I'm asking because I do not have all these different textures to try a performance test, and I guess using always the same ones but in multiple instances could involve caching and so not being reliable as a test.

Thanks for any answer!

**Comment by Bletraut** [comment]
If the texture is already loaded into memory, then when you call Content.Load<Texture2D>(assetName) again, it will not load it again, but will return the same texture.

**Reply by KenshirouLuke** [reply]
Thanks, but I know it; I'm in fact referring to its retrieval from the Content's dictionary after it's loaded.

**Reply by Bletraut** [reply]
ContentManager already has a dictionary for loaded assets.

**Reply by KenshirouLuke** [reply]
I know that too, as written in the post, code included.

**Reply by Bletraut** [reply]
Sorry, I didn't read the post carefully. The string is hashed to find the element and then the typecast is checked. This is probably not the place that could become a bottleneck. The game logic and rendering will take up 90% of the frame.

**Reply by Bletraut** [reply]
But you can implement your own resource manager, for my project I implemented my own resource manager that uses WeakReference to unload resources when they are not needed (for example when loading a new scene).

**Reply by KenshirouLuke** [reply]
Thanks for the reply, yes it's probably not, it was just a projection on a big game with lot of assets (not so rare), and so an observation about a possible performance improvement on the ContentManager class itself.

Anyway, can you tell something about your implementation that involves the WeakReference?
At the moment I'm just Loading the resources basing on their usage, if are general like fonts I load them in the LoadContent phase of the game and I keep the until the end, and if are specific screen-related I load them with a specific instance of the ContentManager owned by the screen, and then unload them when the screen is gone (like, level/scene finished).

**Reply by Bletraut** [reply]
The main idea is to make a wrapper for resources via a [WeakReference](https://learn.microsoft.com/en-us/dotnet/api/system.weakreference?view=net-7.0). AssetManager (my own, internally it still uses ContentManager to load resources) listens to special events, for example, loading a new scene. Then it goes through all loaded assets and looks at the state of the WeakReference. If there are no links, it unloads the resource from memory. Resource checking is done after loading the scene, because the new scene can use the resources of the previous one. This approach is more automated, there is no need to remember which resources are needed for each screen.

**Reply by KenshirouLuke** [reply]
Interesting, thanks for sharing, so if I understand correctly, you load the resources for one or more scenes (and you need to know exactly what resource they need), then when you need to dismiss one or more of them in favore of one or more others, you dont unload the resources with each dismissed scene, but you wait to load the new one(s), and only then you see for resources not needed anymore, in case some of them are shared between different scenes, to avoid unload and reload them. I got it right?

**Reply by Bletraut** [reply]
Yep, you are right. But a small remark, the scene does not know what resources it needs, the components load the resources they need as needed (for example, SpriteRenderer component (my own)). In other words, when the scene is loaded, it loads all the objects, the objects load all the components, and the components load the resources they need. After everything has loaded, the scene is considered loaded. Perhaps during the game, some component will appear that needs another resource and then it will be loaded as needed (there may be a lag if the resource is heavy) in such cases, I put a special component on the scene that will immediately preload the necessary resources.

**Comment by ThomasFOG** [comment]
This really depends on the use-case and if you feel like it is a problem to you.

My philosophy on early optimization is "don't make it a problem unless it is a problem (or you're having fun solving it)".

And if something is a problem, just implement and compare the two approaches with a profiler.

**Reply by KelsamGames** [reply]
^ this.

**Comment by KenshirouLuke** [comment]
Thanks for the answer, yes indeed the string is hashed into an int, let's say this is not all about early optimization logic, even that since I'm a programmer by profession I must analyze immediately the performance impact, to avoid a total refactoring on a second stage.

It's more than an observation about the Monogame framewotk itself: why we have a single dictionary, instead of, let's say, a dictionary of dictionaries, that act as a container, having as key the resource type and as value the related dictionary?
In this way, the first check would be on the type, retrieving immediately the needed dictionary (Texture2D, SpriteFont, Song, etc..) and then the needed item, among few items so faster, and most of all without the object cast:
`return (T)value;`
Expecially since the resources, in Monogame, for some reasons (that I don't know) are not inherited by the same class or by a class at all (except Textures).
Just my two bits!
